Fix: Add Session ID and Header Support to SSE Transport (#660)

naji247 opened the pull request:
# Overview
This is a PR to fix issue #660. The SSE transport in MCP was missing critical request context that StreamableHTTP transport already provided. When using SSE, MCP servers couldn't access:
- Session IDs for connection tracking
- Authentication information from middleware
- Request headers for custom metadata or authorization

This created a feature gap between the two transports, limiting SSE-based implementations. Even though SSE is technically deprecated, our analysis of major companies with MCP servers shows that many are still dependent on SSE. To disable SSE support for their users would require all of them to update their MCP JSON configs (which is quite a hassle). 

This PR brings full feature parity to the SSE transport in the following ways:

1. Session ID Support

- Updated McpSSETransport to accept and store session IDs
- Modified transport initialization to pass session ID from the agent name

2. Authentication & Header Forwarding

- Created buildMessageExtraInfo() helper to construct MessageExtraInfo with auth and headers
- Updated onSSEMcpMessage() to accept and forward authentication context
- Modified SSE handler in utils.ts to extract and pass request metadata

# Key Changes
- McpSSETransport: Now accepts optional sessionId in constructor
- onSSEMcpMessage: Accepts extraInfo parameter with headers and auth
- buildMessageExtraInfo: New helper that filters internal headers and builds proper context
- Type definitions: Added RequestWithAuth type for cleaner code

# Testing
I added a few tests to the existing SSE based test file. The build passes successfully and the changes have been validated to ensure:
- Session IDs are properly transmitted
- Authentication info from middleware is accessible
- Request headers are forwarded correctly
- Internal transport headers are filtered out

naji247 commented:
> this looks good to me, but I'll let @deathbyknowledge have a look before we land it since he'd touched this part of the code

Thank you @threepointone ! Please let us know if you want any modifications. I tried to keep it similar to how Streamable HTTP was implemented.

deathbyknowledge commented on packages/agents/src/tests/mcp/transports/streamable-http.test.ts:
```diff
@@ -583,4 +583,104 @@ describe("Streamable HTTP Transport", () => {
+        method: "POST"
+      });
+
+      // Spy on the transport to capture the extraInfo
```
Doesn't seem like this is testing anything new here. What am I missing?

naji247 commented:
@deathbyknowledge I went ahead and updated the tests, reverted some of the changes I made to include auth information, and refactored the SSE constructor to use an `options` object paradigm.

deathbyknowledge commented on packages/agents/src/mcp/utils.ts:
```diff
@@ -682,7 +682,9 @@ export const createLegacySseHandler = (
 
       const messageBody = await request.json();
-      const error = await agent.onSSEMcpMessage(sessionId, messageBody);
+      const error = await agent.onSSEMcpMessage(sessionId, messageBody, {
```
If we remove `buildMessageExtraInfo` this can be:
```ts
      const headers =  Object.fromEntries(request.headers.entries());
      const error = await agent.onSSEMcpMessage(sessionId, messageBody, {
        requestInfo: { headers }
      });
```

deathbyknowledge reviewed (changes requested):
Looks good! There are a few nits and one type error breaking the CI but this should be ready to merge after those are addressed 👍

naji247 commented:
@deathbyknowledge should be good to go :)

deathbyknowledge commented:
CI is failing, seems like a silly type error. Can you check with `npm run check` please?
